man/audio2tape.1
.\" -*- nroff -*-
.\"
.\" audio2tape.1: audio2tape man page
.\" Copyright (c) 2007-2008 Fredrick Meunier
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.\"
.\" Author contact information:
.\"
.\" E-mail: fredm@spamcop.net
.\"
.\"
.TH audio2tape 1 "24th May, 2013" "Version 1.1.1" "Emulators"
.\"
.\"------------------------------------------------------------------
.\"
.SH NAME
audio2tape \(em Sinclair ZX Spectrum audio to tape file converter
.\"
.\"------------------------------------------------------------------
.\"
.SH SYNOPSIS
.B audio2tape
.RI "[\-t " type ]
[\-s]
[\-r]
[\-k]
.RI "[\-z " level ]
.RI "[\-c " level ]
.I infile outfile
.\"
.\"------------------------------------------------------------------
.\"
.SH DESCRIPTION
audio2tape converts audio files to ZX Spectrum tape images.
.\"
.\"------------------------------------------------------------------
.\"
.SH OPTIONS
.TP
.RI "\-t " type
causes
.I type
to be used as the level detection trigger. There are two triggers implemented;
the
.I simple
trigger just uses a threshold value to convert to a one or zero on output, and
the
.I schmitt
trigger implements a Schmitt trigger which outputs a less noisy signal, but
requires a stronger signal to noise ratio than the simple trigger. The default
trigger is \(lqschmitt\(rq.
.TP
\-s
causes the display of a frequency graph showing the sound levels present in the
audio file. This can be useful when selecting good crossover points.
.TP
\-r
causes any ROM blocks that are recognised to use idealised timing figures
rather than the exact ones found in the audio file.
.TP
\-k
causes the tape file that is produced to keep any audio data from the original
audio file that isn't recognised. This will allow unsupported loading schemes,
encryption tones etc. to be preserved in the file that is produced.
.TP
.RI "\-z " level
causes
.I level
to be used as the zero point in the supplied audio file, it allows adjustment
for a DC offset in the audio file. The default level is 127 from a range of
0\(en255.
.TP
.RI "\-c " level
causes
.I level
to be used as the Schmitt trigger noise threshold, this reduces the sensitivity
to noise in the provided file. The default level is 8 from a range of 0\(en127.
.TP
.I infile
specifies the audio file to be converted. This file can be in any of the
sound formats supported by
.IR libaudiofile .
.TP
.I outfile
specifies the output tape file. The format will be determined by the
file extension and can be in any of the formats
.IR libspectrum "(3)"
supports for output.
.\"
.\"------------------------------------------------------------------
.\"
.SH BUGS
None known.
.\"
.\"------------------------------------------------------------------
.\"
.SH SEE ALSO
.IR fuse "(1),"
.IR fuse\-utils "(1),"
.IR libspectrum "(3),"
.IR listbasic "(1),"
.IR tapeconv "(1),"
.IR tzxlist "(1),"
.IR snap2tzx "(1),"
.IR tape2wav "(1)"
.PP
The comp.sys.sinclair Spectrum FAQ, at
.br
.IR "http://www.worldofspectrum.org/faq/index.html" .
.\"
.\"------------------------------------------------------------------
.\"
.SH AUTHOR
Fredrick Meunier (fredm@spamcop.net).


ChangeLog
2013-05-xx  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

        * libspectrum 1.1.1 released.

        * Decreased libtool version to allow binaries built against
          libspectrum 1.0.0 to run with libspectrum 1.1.1.

2013-05-18  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

        * libspectrum 1.1.0 released.

        * Support for Spectranet in snapshots (Philip Kendall).

        * Support for Cheetah SpecDrum in snapshots (Fredrick Meunier).

        * Support reading PZX files (Fredrick Meunier).

        * Support for the TZX set signal level block (Fredrick Meunier).

        * Various minor bug fixes/improvements:
          * Add support for detecting POK files (Sergio Baldoví).
          * Advertise support for bzip2 compressed and wav files (Sergio
            Baldoví).
          * Drop support for GLib version 1.x (Alberto Garcia).
          * Add support for audiofile 0.3.x (Alberto Garcia).
          * Load all snapshot RAM pages for machines with more than 128k 
            (Fredrick Meunier; thanks, ketmar).
          * Support more SZX 1.5 features (Fredrick Meunier).
          * Modernise autoconf support (Sergio Baldoví).

2010-12-16  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

        * libspectrum 1.0.0 released.

        * New end of tape flag (Fredrick Meunier).

        * AF and AF' were swapped when reading and writing SZX files
          (Fredrick Meunier; thanks, velesoft).

        * Support for Fuller Box and Melodik (and similar AY interfaces) in
          snapshots (Fredrick Meunier).

        * Support for Opus Discovery in snapshots (Fredrick Meunier, Gergely
          Szasz).

        * Support for the NTSC 48K Spectrum (Philip Kendall).

        * Support for stereo sound files (Fredrick Meunier; thanks, Gilberto
          Almeida).

        * Add proper support for +2A/+3, Pentagon and Scorpion in .z80 files
          (Philip Kendall; thanks, Nick Corey)

        * Various fixes for compiling with Visual Studio 6 (Simon Owen).

        * Various minor bug fixes/improvements:
          * Looser signature checks for .dsk files (Gergely Szasz; thanks,
            zx81 and Simon Owen).
          * --without-(bzip2|zlib) options to configure for Gentoo and similar
            distributions (José Manuel Ferrer Ortiz).
          * Handle querying the block length for turbo and pure data blocks
            with 0 length (Fredrick Meunier).
          * Fix loading of Warajevo TAPs with raw samples (Fredrick Meunier).
          * Fail sensibly if file decompression fails (Fredrick Meunier).
          * Remove various unused variables, initialisations etc (Fredrick
            Meunier).

2008-11-30  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

        * libspectrum 0.5.0 released.

	* Memory management change: memory allocation functions now abort
	  on failure. Leads to simpler code both internally and for
	  users of the library (Philip Kendall).

	* Support for tape acceleration (Philip Kendall)

	* IDE improvements to allow DivIDE to support FATware (Philip
	  Kendall; thanks, http://www.dusky.sk/zxs/)

	* New API to deal with flagging automatic snapshots in RZX files
	  (Philip Kendall).

	* New function to get length (time) of a tape block (Philip Kendall,
	  Fredrick Meunier).

	* 128K SNA files are now treated as being for the Pentagon 128
	  rather than the Spectrum 128 (Fredrick Meunier)

	* SNA files now written out correctly (Philip Kendall).

	* Handle MDR write protection flag correctly (Philip Kendall)

	* Allow MDR images with bad checksums to load (the error will
	  be handled by the If1 ROM) (Gergely Szasz)

        * Support saving custom ROMs in SZX snapshots (Fredrick Meunier,
	  Stuart Brady)

	* Support for DivIDE, Kempston mouse and the simple IDE interface
	  in SZX snapshots (Fredrick Meunier)

        * Allow building without libgcrypt (Stuart Brady)

	* Amiga and MorphOS compilation fixes (Chris Young, Q-Master)

	* Don't produce empty libraries to fix build process on
	  OpenSolaris (Fredrick Meunier)

	* Remove deprecated functions (Fredrick Meunier).

	* Tweak timings for TS2068 (Fredrick Meunier; thanks Philip Kendall).

        * NTSC Spectrum support (Philip Kendall; thanks, Claudio Bernet).

2008-01-05  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

	* libspectrum 0.4.0 released.

	* Add support for loading from WAV files (Fredrick Meunier).

	* Add support for CSW tape files (Fredrick Meunier).

	* Tweak timings for TC2048 and TC2068 (Fredrick Meunier; thanks Philip
	  Kendall and Mark Woodmass).

	* Add support for the +D in z80 and szx snapshots (Fredrick Meunier).

	* Add Pentagon 512/1024 support (Q-Master).

	* Allow Amiga version to decompress files with xfdmaster.library
	  (Chris Young).

	* Allow MorphOS version to use xfdmaster.library (Q-Master).

	* Various other bugfixes.

2007-05-11  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

	* libspectrum 0.3.0.1 released (bug fix release).

	* Avoid crash when reading an invalid TZX file (thanks, crabfists).
	  (Philip Kendall).

	* Export the TZX Generalized Data Block functions when building a
	  Win32 DLL (Stuart Brady).

	* Distribute the hacking/ directory (thanks, Stuart Brady) (Philip
	  Kendall).

	* Remove warning when building from CVS (Stuart Brady).

2007-04-11  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

	* libspectrum 0.3.0 released.

	* Support for RZX rollback, including having snapshots and
	  multiple input recording blocks in input recording files (Philip
	  Kendall).

	* Improved IDE emulation (Matthew Westcott).

	* Support for microdrive images (Philip Kendall, Gergely Szasz).

	* TZX Generalized Data Block support. (Philip Kendall).

	* Support for Z80Em and CSW tape files (Darren Salt).

	* Support for the TS2068 (Fredrick Meunier).

	* Support for connected joystick information in snapshots (Fredrick
	  Meunier).

	* Add interrupt length to timing information (Philip Kendall).


audio2tape.c
/* audio2tape.c: Convert audio files (wav, voc, etc.) to tape files (.tzx,.csw)
   Copyright (c) 2007 Fredrick Meunier

   $Id$

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License along
   with this program; if not, write to the Free Software Foundation, Inc.,
   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

   Author contact information:

   E-mail: fredm@spamcop.net

*/

#include <config.h>

#include <errno.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>

#include <libspectrum.h>

#include <audiofile.h>

#include "utils.h"

typedef  libspectrum_byte (*glm) ( libspectrum_byte byte );

static int read_tape( glm get_level_method, char *filename, libspectrum_tape **tape );
static int write_tape( char *filename, libspectrum_tape *tape );

static void collect_stats( libspectrum_byte byte );
static void display_stats( void );

char *progname;

int show_stats = 0;

libspectrum_byte simple_crossover_point = 127;

static libspectrum_byte
get_level_simple( libspectrum_byte byte )
{
  return byte > simple_crossover_point;
}

libspectrum_byte schmitt_low2high_crossover_point = 0xc3;
libspectrum_byte schmitt_high2low_crossover_point = 0x3c;

static libspectrum_byte
get_level_schmitt( libspectrum_byte byte )
{
  static libspectrum_byte last_level = 0;

  switch( last_level ) {
  case 0:
    if( byte >= schmitt_low2high_crossover_point )
      last_level = 1;
    break;
  case 1:
    if( byte < schmitt_high2low_crossover_point )
      last_level = 0;
    break;
  }

  return last_level;
}

const char *schmitt = "schmitt";
const char *simple = "simple";

static int
set_trigger_type_from_string( glm* get_level_method, const char *string )
{
  if( strncmp( string, simple, strlen(simple) ) ) {
    *get_level_method = get_level_simple;
  } else if( strncmp( string, schmitt, strlen(schmitt) ) ) {
    *get_level_method = get_level_schmitt;
  } else {
    fprintf( stderr, "%s: unrecognised trigger type: %s\n", progname, string );
    return 1;
  }
  
  return 0;
}

int
main( int argc, char **argv )
{
  int c, error;
  char *trigger_type_string = NULL; glm get_level_method;
  libspectrum_tape *tzx;

  progname = argv[0];

  error = init_libspectrum(); if( error ) return error;

  get_level_method = get_level_simple;

  while( ( c = getopt( argc, argv, "t:sp:h:l:" ) ) != -1 ) {

    switch( c ) {

    case 't': trigger_type_string = optarg; break;
    case 's': show_stats = 1; break;
    case 'p': simple_crossover_point = atol( optarg ); break;
    case 'h': schmitt_low2high_crossover_point = atol( optarg ); break;
    case 'l': schmitt_high2low_crossover_point = atol( optarg ); break;

    }
  }

  get_level_method = get_level_simple;
  if( trigger_type_string &&
      set_trigger_type_from_string( &get_level_method, trigger_type_string ) )
    return 1;

  argc -= optind;
  argv += optind;

  if( argc < 2 ) {
    fprintf( stderr,
             "%s: usage: %s [-t <trigger type>] <infile> <outfile>\n",
             progname,
	     progname );
    return 1;
  }

  if( read_tape( get_level_method, argv[0], &tzx ) ) return 1;

  if( write_tape( argv[1], tzx ) ) {
    libspectrum_tape_free( tzx );
    return 1;
  }

  libspectrum_tape_free( tzx );

  return 0;
}

unsigned int amplitude_frequency_table[256];

static void
collect_stats( libspectrum_byte byte )
{
  amplitude_frequency_table[byte]++;
}

static void
display_stats( void )
{
  int i, j;
  printf("Frequency table:\n");
  for( i = 0; i < 32; i++ ) {
    int subtotal = 0;
    for( j = 0; j < 8; j++ ) {
      subtotal += amplitude_frequency_table[i*8+j];
    }
    printf( "%3d-%3d | %d\n", i*8, i*8+7, subtotal );
  }
}

static int
read_tape( glm get_level_method, char *filename, libspectrum_tape **tape )
{
  libspectrum_byte *buffer; size_t length;
  libspectrum_byte *tape_buffer; size_t tape_length;

  /* Our filehandle from libaudiofile */
  AFfilehandle handle;

  /* The track we're using in the file */
  int track = AF_DEFAULT_TRACK; 

  *tape = libspectrum_tape_alloc();

  handle = afOpenFile( filename, "r", NULL );
  if( handle == AF_NULL_FILEHANDLE ) {
    libspectrum_tape_free( *tape );
    return 1;
  }

  if( afSetVirtualSampleFormat( handle, track, AF_SAMPFMT_UNSIGNED, 8 ) ) {
    libspectrum_tape_free( *tape );
    afCloseFile( handle );
    return 1;
  }

  length = afGetFrameCount( handle, track );

  tape_length = length;
  if( tape_length%8 ) tape_length += 8 - (tape_length%8);

  buffer = calloc( tape_length, 1);
  if( !buffer ) {
    libspectrum_tape_free( *tape );
    afCloseFile( handle );
    return 1;
  }

  int frames = afReadFrames( handle, track, buffer, length );
  if( frames == -1 ) {
    fprintf( stderr, "Can't calculate number of frames\n" );
    free( buffer );
    libspectrum_tape_free( *tape );
    afCloseFile( handle );
    return 1;
  }

  if( !length ) {
    fprintf( stderr, "Empty audiofile, nothing to convert\n" );
    free( buffer );
    libspectrum_tape_free( *tape );
    afCloseFile( handle );
    return 1;
  }

  if( frames != length ) {
    fprintf( stderr, "Read %d frames, but expected %lu\n", frames,
	     (unsigned long)length );
    free( buffer );
    libspectrum_tape_free( *tape );
    afCloseFile( handle );
    return 1;
  }

  libspectrum_tape_block *block = 
    libspectrum_tape_block_alloc( LIBSPECTRUM_TAPE_BLOCK_RAW_DATA );

  /* 44100 Hz 79 t-states 22050 Hz 158 t-states */
  libspectrum_tape_block_set_bit_length( block,
                                         3500000/afGetRate( handle, track ) );
  libspectrum_tape_block_set_pause     ( block, 0 );
  libspectrum_tape_block_set_bits_in_last_byte( block, length%8 ? length%8 : 8 );
  libspectrum_tape_block_set_data_length( block, tape_length/8 );

  tape_buffer = calloc( tape_length/8, 1 );

  libspectrum_byte *from = buffer;
  libspectrum_byte *to = tape_buffer;
  length = tape_length;
  do {
    libspectrum_byte val = 0;
    int i;
    for( i = 7; i >= 0; i-- ) {
      collect_stats( *from );
      if( get_level_method( *from++ ) ) val |= 1 << i;
    }
    *to++ = val;
  } while ((length -= 8) > 0);

  if( show_stats ) display_stats();

  libspectrum_tape_block_set_data( block, tape_buffer );

  /* Finally, put the block into the block list */
  libspectrum_tape_append_block( *tape, block );

  if( afCloseFile( handle ) ) {
    free( buffer );
    return 1;
  }

  free( buffer );

  return 0;
}

static int
write_tape( char *filename, libspectrum_tape *tape )
{
  libspectrum_byte *buffer; size_t length;
  FILE *f;
  libspectrum_id_t type;
  libspectrum_class_t class;
  int error;

  /* Work out what sort of file we want from the filename; default to
     .tzx if we couldn't guess */
  error = libspectrum_identify_file_with_class( &type, &class, filename, NULL,
						0 );
  if( error ) return error;

  if( class != LIBSPECTRUM_CLASS_TAPE || type == LIBSPECTRUM_ID_UNKNOWN )
    type = LIBSPECTRUM_ID_TAPE_TZX;

  length = 0;

  if( libspectrum_tape_write( &buffer, &length, tape, type ) ) return 1;

  f = fopen( filename, "wb" );
  if( !f ) {
    fprintf( stderr, "%s: couldn't open '%s': %s\n", progname, filename,
	     strerror( errno ) );
    free( buffer );
    return 1;
  }
    
  if( fwrite( buffer, 1, length, f ) != length ) {
    fprintf( stderr, "%s: error writing to '%s'\n", progname, filename );
    free( buffer );
    fclose( f );
    return 1;
  }

  free( buffer );

  if( fclose( f ) ) {
    fprintf( stderr, "%s: couldn't close '%s': %s\n", progname, filename,
	     strerror( errno ) );
    return 1;
  }

  return 0;
}
